plugins/slack_app/app/controllers/slack_app_controller.rb

```diff
@@ -0,0 +1,152 @@
+    else
+      SlackIdentifier.find_by_identifier(params[:user_id]).try(:user)
+    end
+  end
+
+  def oauth
+    @need_to_connect_app = @app_token.nil?
+    @need_to_connect_user = @current_user_from_slack.nil?
+
+    if params[:code]
+      # Got an OAuth code, let's fetch our tokens
+      resp = slack.oauth_access client_id: ENV['SLACK_CLIENT_ID'],
+                                client_secret: ENV['SLACK_CLIENT_SECRET'],
+                                code: params[:code],
+                                redirect_uri: url_for(only_path: false)
+      if @need_to_connect_app
+        @app_token = SlackIdentifier.create!(identifier: resp['access_token']).identifier
+      end
+      SlackIdentifier.create!(user_id: current_user.id, identifier: resp['user_id'])
+
+      redirect_to
+    end
+  end
+
+  def command
+    render json: command_response
+  end
+
+  def interact
+    # Interesting parts of the payload:
+    # {
+    #   "actions"=>[
+    #     {"name"=>"one", "value"=>"one"} <-- which button the user clicked
+    #   ],
+    #   "callback_id"=>"deploy-123456", <-- maps to a deployment
+    #   "user"=>{"id"=>"U0HAGH3AB", "name"=>"ben"}, <-- who clicked
+    #   "action_ts"=>"1468259216.482422",
+    #   "message_ts"=>"1468259019.000006", <-- need this to maybe update the message
+    #   "response_url"=>"https://hooks.slack.com/actions/T0HAGP0J2/58691408951/iHdjN2zjX0dZ5rXyl84Otql7"
+    # }
+    render json: interact_response # This replaces the original message
+  end
+
+  private
+
+  def command_response
+    # Interesting parts of params:
+    # {
+    #   "user_id"=>"U0HAGH3AB", <-- look up Samson user using this
+    #   "command"=>"/deploy",
+    #   "text"=>"foo bar baz", <-- this is what to deploy
+    #   "response_url"=>"https://hooks.slack.com/commands/T0HAGP0J2/58604540277/g0xd4K2KOsgL9zXwR4kEc0eL"
+    #    ^^^^ This is how to respond later, we can also directly render some JSON
+    # }
+    return unknown_user unless @current_user_from_slack
+
+    # Parse the command
+    projectname, branchname, stagename = @payload['text'].match(/([^\s\/]+)(?:\/(\S+))?\s*(?:to\s+(.*))?/).captures
+
+    # Sanity checks
+    begin
+      project = Project.find_by_param! projectname
+    rescue ActiveRecord::RecordNotFound
+      return unknown_project(projectname)
+    end
+
+    return unauthorized_deployer unless @current_user_from_slack.deployer_for?(project)
+
+    begin
+      stage = project.stages.find_by_param!(stagename || 'production')
+    rescue ActiveRecord::RecordNotFound
+      return unknown_stage(project, stagename)
+    end
+
+    deploy_service = DeployService.new(@current_user_from_slack)
+    deploy = deploy_service.deploy!(stage, reference: branchname || 'master')
+    DeployResponseUrl.create! deploy: deploy, response_url: @payload[:response_url]
+
+    SlackMessage.new(deploy).message_body
+  end
+
+  def interact_response
+    # This means someone clicked the "Approve" button in the channel
```

Comment:
This isn't cleared with compliance for use in our instance yet, but I think people wanted to move forward with the PR.  Before we enable the plugin for our instance it'll have to be cleared with compliance.  Having the PR cleaned up will make it easier to reviews when we go through and clear it or add some security requirements.

> On Jul 17, 2016, at 09:02, Michael Grosser <notifications@github.com> wrote:
> 
> In plugins/slack_app/app/controllers/slack_app_controller.rb:
> 
> > +
> > +    begin
> > +      stage = project.stages.find_by_param!(stagename || 'production')
> > +    rescue ActiveRecord::RecordNotFound
> > +      return unknown_stage(project, stagename)
> > +    end
> > +
> > +    deploy_service = DeployService.new(@current_user_from_slack)
> > +    deploy = deploy_service.deploy!(stage, reference: branchname || 'master')
> > +    DeployResponseUrl.create! deploy: deploy, response_url: @payload[:response_url]
> > +
> > +    SlackMessage.new(deploy).message_body
> > +  end
> > +
> > +  def interact_response
> > +    # This means someone clicked the "Approve" button in the channel
> maybe I lost track of the conversation but never heard that we cleared this with compliance ?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>